tidy3d/components/simulation.py

```diff
@@ -439,6 +439,71 @@ def diffraction_monitor_boundaries(cls, val, values):
                         )
         return val
 
+    # pylint: disable=too-many-locals
+    @pydantic.validator("monitors", always=True)
+    def _projection_monitors_homogeneous(cls, val, values):
+        """Error if any field projection monitor is not in a homogeneous region."""
+
+        if val is None:
+            return val
+
+        # list of structures including background as a Box()
+        structure_bg = Structure(
+            geometry=Box(
+                size=values.get("size"),
+                center=values.get("center"),
+            ),
+            medium=values.get("medium"),
+        )
+
+        structures = values.get("structures") or []
+        total_structures = [structure_bg] + list(structures)
+
+        # in case of a volume monitor, we only care about the surfaces, so create a list of
+        # surface monitors
+        monitors = []
+        for monitor in val:
+            if isinstance(monitor, (AbstractNear2FarMonitor, DiffractionMonitor)):
+                if monitor.size.count(0.0) == 0:
+                    monitor_dict = monitor.dict()
+                    monitor_dict.pop("exclude_surfaces", None)
+                    surface_monitors = monitor.surfaces(**monitor_dict)
+                    exclude_surfaces = monitor.exclude_surfaces
+                    if exclude_surfaces:
+                        surface_monitors = [
+                            mnt for mnt in surface_monitors if mnt.name[-2:] not in exclude_surfaces
+                        ]
+                    monitors += surface_monitors
+                else:
+                    monitors.append(monitor)
+
+        # now we have a list of field projection surface monitors
+        for monitor in monitors:
+            mediums = cls.intersecting_media(monitor, total_structures)
+            # make sure there is no more than one medium in the returned list
+            if len(mediums) > 1:
+                raise SetupError(
+                    f"{len(mediums)} different mediums detected on plane "
+                    f"intersecting a {monitor.type}. Plane must be homogeneous."
+                )
```

Comment:
Isn't this now redundant as it is already done in `Simulation.intersecting_media`?

Reply:
The difference is in how `exclude_surfaces` is treated. `intersecting_media` is a generic function that doesn't care about `exclude_surfaces` (except to `pop` it so that a validator isn't accidentally triggered). But we may have a case where a near2far box's surface intersects more than one medium, but that surface is also excluded, in which case we won't trigger the homogeneous medium validator.

Reply:
It's not very clean, I agree. We could also take `exclude_surfaces` into account directly in `intersecting_media`.

Reply:
Oh I see. Then the issue is not just cleanness though. In `monitor_medium`, the error check will be triggered if the monitor spans more than one medium even if the surfaces in the second medium are correctly "excluded".

Whichever approach you take in fixing this, there should be a test like that.

Reply:
Ah yeah I see what you mean

Reply:
Ok, I've made `intersecting_media` take `exclude_surfaces` into account, and added a test.